File: ios/Sources/Omnesis/UI/PushHealthBanner.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

/// Surfaces the ways this phone can stop delivering data to the gateway
/// while every other indicator still reads "synced".
///
/// The app's sync status answers "did the collector run?", which is not the
/// same question as "did the data arrive?". A sync cycle reads from HealthKit,
/// writes batches to the offline buffer, and advances its cursor there — the
/// upload is a separate step. When that step fails, the source keeps reporting
/// completed syncs and the cursor keeps moving, so the phone looks healthy
/// while the gateway silently goes stale.
///
/// Three distinct failures get their own copy because the remedies differ:
///
///   - **Blocked** — the gateway refused a source with 403 because this
///     device's token lacks its `write:<source-type>` scope. Nothing the user
///     can fix on the phone; it needs a grant on the gateway. The buffered data
///     is intact and delivers as soon as the scope lands.
///   - **Backlog** — batches are queued and the oldest one keeps aging. The
///     link is down, the gateway is unreachable, or uploads are erroring. Retry
///     is the right action.
///   - **Undelivered** — the uploader gave up on some batches and set them
///     aside (see `OfflineBuffer`'s quarantine). Nothing is pending any more,
///     so there is nothing to retry; what is left is to say the data didn't
///     make it and let the user clear the notice.
///
/// A merely non-empty buffer is *not* a problem — that's the normal state
/// mid-sync — so the backlog case is gated on the oldest entry's age rather
/// than on the count.
///
/// Retry reports back: the outcome (`DrainOutcome`) is rendered, including
/// the one that matters most here — "the gateway answered and refused",
/// which no amount of tapping will change. The phase is supplied by the
/// caller rather than held here; see `PushHealth.RetryPhase`.
///
/// The decisions this view renders — when a state is worth reporting, and
/// what each outcome says — live in `PushHealth`.
@available(iOS 17.0, *)
struct PushHealthBanner: View {
    let blockedSourceIds: [String]
    let bufferedBatches: Int
    let oldestBufferedAge: TimeInterval?
    /// Batches the uploader set aside as undeliverable.
    let quarantinedBatches: Int
    /// Resolves a source id to the name to show. The banner never spells out a
    /// source name of its own (see the source-encapsulation rule in CLAUDE.md)
    /// — the caller resolves through the descriptor registry.
    let labelForSourceId: (String) -> String
    let retryPhase: PushHealth.RetryPhase
    let onRetry: () -> Void
    let onDiscardUndelivered: () -> Void

    @State private var confirmingDiscard = false

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            if !blockedSourceIds.isEmpty {
                blockedRow
            }
            if PushHealth.isBacklogged(oldestBufferedAge, blockedSourceIds: blockedSourceIds) {
                backlogRow
            }
            if quarantinedBatches > 0 {
                undeliveredRow
            }
        }
        .padding(.vertical, 4)
    }

    private var blockedRow: some View {
        VStack(alignment: .leading, spacing: 6) {
            Label("Omnesis isn't accepting some data", systemImage: "lock.trianglebadge.exclamationmark.fill")
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(Theme.danger)
            Text(blockedNames)
                .font(.footnote.weight(.medium))
                .foregroundStyle(Theme.textPrimary)
            Text(
                "This phone isn't allowed to send \(blockedSourceIds.count == 1 ? "this source" : "these sources") yet. "
                    + "The data is held on the phone and uploads once the gateway grants permission — but the "
                    + "phone only keeps a limited backlog, so the oldest of it is dropped if this isn't fixed."
            )
            .font(.footnote)
            .foregroundStyle(Theme.textMuted)
            .fixedSize(horizontal: false, vertical: true)
        }
        .accessibilityElement(children: .combine)
    }

    private var backlogRow: some View {
        VStack(alignment: .leading, spacing: 6) {
            Label("Data isn't reaching Omnesis", systemImage: "exclamationmark.arrow.trianglehead.2.clockwise.rotate.90")
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(Theme.warning)
            Text(backlogDetail)
                .font(.footnote)
                .foregroundStyle(Theme.textMuted)
                .fixedSize(horizontal: false, vertical: true)
            retryControl
            if case .reported(let outcome) = retryPhase {
                RetryStatusLine(outcome: outcome)
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Delivery backlog")
    }

    @ViewBuilder
    private var retryControl: some View {
        if retryPhase == .running {
            HStack(spacing: 6) {
                ProgressView().controlSize(.small)
                Text("Uploading…")
                    .font(.footnote.weight(.semibold))
                    .foregroundStyle(Theme.textMuted)
            }
        } else {
            Button(retryPhase == .idle ? "Retry now" : "Try again", action: onRetry)
                .font(.footnote.weight(.semibold))
                .buttonStyle(.borderless)
        }
    }

    private var undeliveredRow: some View {
        VStack(alignment: .leading, spacing: 6) {
            Label("Some data never reached Omnesis", systemImage: "tray.full")
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(Theme.textPrimary)
            Text(
                "\(Self.batchCount(quarantinedBatches)) couldn't be delivered after repeated attempts, so "
                    + "\(quarantinedBatches == 1 ? "it was" : "they were") set aside. Everything since is syncing "
                    + "normally. Set-aside data is kept on this phone for a while, and discarding it now frees "
                    + "the space immediately."
            )
            .font(.footnote)
            .foregroundStyle(Theme.textMuted)
            .fixedSize(horizontal: false, vertical: true)
            Button("Discard", role: .destructive) { confirmingDiscard = true }
                .font(.footnote.weight(.semibold))
                .buttonStyle(.borderless)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Undelivered data")
        .confirmationDialog(
            "Discard undelivered data?",
            isPresented: $confirmingDiscard,
            titleVisibility: .visible
        ) {
            Button("Discard", role: .destructive, action: onDiscardUndelivered)
            Button("Keep", role: .cancel) {}
        } message: {
            Text("\(Self.batchCount(quarantinedBatches)) will be deleted from this phone. This can't be undone.")
        }
    }

    private var blockedNames: String {
        blockedSourceIds
            .map(labelForSourceId)
            .formatted(.list(type: .and))
    }

    private static func batchCount(_ count: Int) -> String {
        count == 1 ? "1 batch" : "\(count) batches"
    }

    private var backlogDetail: String {
        let batches = Self.batchCount(bufferedBatches)
        guard let oldestBufferedAge else {
            return "\(batches) waiting to upload."
        }
        // A duration ("3 days"), not a relative date ("3 days ago") — the point
        // is how long the data has been stuck, not when it was captured.
        let waited = Duration.seconds(oldestBufferedAge)
            .formatted(.units(allowed: [.days, .hours, .minutes], maximumUnitCount: 1))
        return "\(batches) waiting to upload — the oldest has been waiting \(waited)."
    }
}

/// The one line a finished retry leaves behind. Its own view so a snapshot
/// can render every outcome side by side, which the banner itself only ever
/// shows one at a time.
@available(iOS 17.0, *)
struct RetryStatusLine: View {
    let outcome: DrainOutcome

    var body: some View {
        Text(PushHealth.retryMessage(for: outcome))
            .font(.footnote)
            .foregroundStyle(PushHealth.retryIsTrouble(outcome) ? Theme.warning : Theme.textMuted)
            .fixedSize(horizontal: false, vertical: true)
    }
}

#if DEBUG
/// Wraps a banner in the Form chrome both call sites give it, so previews
/// differ only in the state under test.
@available(iOS 17.0, *)
private struct PushHealthPreview: View {
    var blockedSourceIds: [String] = []
    var bufferedBatches: Int = 0
    var oldestBufferedAge: TimeInterval?
    var quarantinedBatches: Int = 0
    var retryPhase: PushHealth.RetryPhase = .idle

    var body: some View {
        Form {
            Section {
                PushHealthBanner(
                    blockedSourceIds: blockedSourceIds,
                    bufferedBatches: bufferedBatches,
                    oldestBufferedAge: oldestBufferedAge,
                    quarantinedBatches: quarantinedBatches,
                    labelForSourceId: { _ in "Places" },
                    retryPhase: retryPhase,
                    onRetry: {},
                    onDiscardUndelivered: {}
                )
            }
            .listRowBackground(Theme.bgSecondary)
        }
        .scrollContentBackground(.hidden)
        .background(Theme.bgPrimary)
    }
}

#Preview("Blocked — one source") {
    PushHealthPreview(
        blockedSourceIds: ["core-location-visits:local"],
        bufferedBatches: 12,
        oldestBufferedAge: 60 * 60
    )
}

#Preview("Blocked — several sources") {
    PushHealthPreview(
        blockedSourceIds: ["core-location-visits:local", "photos:local"],
        bufferedBatches: 340,
        oldestBufferedAge: 3 * 24 * 60 * 60
    )
}

#Preview("Backlog only") {
    PushHealthPreview(bufferedBatches: 87, oldestBufferedAge: 3 * 24 * 60 * 60)
}

#Preview("Blocked with an empty queue") {
    PushHealthPreview(blockedSourceIds: ["core-location-visits:local"], bufferedBatches: 1)
}

#Preview("Undelivered only") {
    PushHealthPreview(quarantinedBatches: 3)
}

#Preview("Undelivered — a single batch") {
    PushHealthPreview(quarantinedBatches: 1)
}

#Preview("Backlog — retry in flight") {
    PushHealthPreview(
        bufferedBatches: 87,
        oldestBufferedAge: 3 * 24 * 60 * 60,
        retryPhase: .running
    )
}

#Preview("Backlog — retry refused") {
    PushHealthPreview(
        bufferedBatches: 87,
        oldestBufferedAge: 3 * 24 * 60 * 60,
        retryPhase: .reported(.refused)
    )
}

#Preview("Backlog and undelivered together") {
    PushHealthPreview(
        bufferedBatches: 9,
        oldestBufferedAge: 12 * 60 * 60,
        quarantinedBatches: 2
    )
}

// A blocked source and batches already given up on. The backlog row stays
// away even though the queue is stale: a blocked source's batches sit at the
// head of the FIFO and age without bound, so reporting them would raise a
// second alarm whose Retry cannot help.
#Preview("Blocked and undelivered") {
    PushHealthPreview(
        blockedSourceIds: ["photos:local"],
        bufferedBatches: 41,
        oldestBufferedAge: 2 * 24 * 60 * 60,
        quarantinedBatches: 5
    )
}

// Every message a finished retry can leave, together — the nine strings are
// only ever seen one at a time in the app, and this is where their length
// and wrapping can be compared.
#Preview("Retry outcomes") {
    Form {
        Section {
            VStack(alignment: .leading, spacing: 14) {
                ForEach(
                    [
                        DrainOutcome.delivered, .refused, .paused, .blocked,
                        .unreachable, .stalled, .busy, .idle, .failed,
                    ],
                    id: \.self
                ) { outcome in
                    RetryStatusLine(outcome: outcome)
                }
            }
        }
        .listRowBackground(Theme.bgSecondary)
    }
    .scrollContentBackground(.hidden)
    .background(Theme.bgPrimary)
}
#endif

#endif

File: android/app/src/main/kotlin/dev/omnesis/android/ui/common/PushHealthBanner.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.common

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Inbox
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.SyncProblem
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.omnesis.android.designsystem.components.OmSpinner
import dev.omnesis.android.designsystem.theme.OmSpacing
import dev.omnesis.android.designsystem.theme.OmTheme
import dev.omnesis.android.designsystem.theme.OmnesisTheme
import dev.omnesis.android.transport.PushHealth
import dev.omnesis.android.transport.PushHealthSnapshot
import dev.omnesis.android.transport.RetryOutcome
import dev.omnesis.android.transport.SetAside
import dev.omnesis.android.transport.SkippedPush

/**
 * Surfaces the ways this phone can stop delivering data while every other
 * indicator still reads "synced".
 *
 * A sync answers "did the source run?", which is not the same question as "did
 * the data arrive?". A pass reads from Android, pushes, and advances a cursor —
 * and the push is the part that fails on its own. When it does, the source
 * keeps reporting completed syncs and the cursor keeps moving, so the phone
 * looks healthy while the gateway silently goes stale.
 *
 * Three failures get their own row because the remedies differ:
 *
 *  - **Blocked** — the gateway refused this device's authority to send for a
 *    source. Nothing on the phone fixes it; it needs a grant on the gateway,
 *    or a fresh pairing.
 *  - **Backlog** — a source that owns a durable queue has one, and its oldest
 *    row keeps aging. Retry is the right action.
 *  - **Undelivered** — a source stopped re-sending something. Nothing is
 *    pending any more, so there is nothing to retry; what is left is to say
 *    the data didn't make it, and to let the user clear the notice.
 *
 * A merely non-empty queue is *not* a problem — that is the normal state
 * mid-sync — so the backlog row is gated on the oldest row's age, never on the
 * count.
 *
 * Every source name and unit noun is resolved by the caller through the source
 * catalog: this view never spells one out.
 */
@Composable
fun PushHealthBanner(
    snapshot: PushHealthSnapshot,
    retryPhase: PushHealth.RetryPhase,
    labelForSourceId: (String) -> String,
    unitLabelForSourceId: (String, Int) -> String,
    onRetry: () -> Unit,
    onDiscardUndelivered: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // Nothing worth reporting renders nothing at all — no empty row, and no
    // padding from a modifier the caller supplied for the populated case.
    if (PushHealth.isHealthy(snapshot)) return
    var confirmingDiscard by remember { mutableStateOf(false) }
    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(OmSpacing.md)) {
        if (snapshot.blockedSourceIds.isNotEmpty()) {
            BlockedRow(snapshot.blockedSourceIds, labelForSourceId)
        }
        if (PushHealth.isBacklogged(snapshot)) {
            BacklogRow(
                snapshot = snapshot,
                retryPhase = retryPhase,
                labelForSourceId = labelForSourceId,
                unitLabelForSourceId = unitLabelForSourceId,
                onRetry = onRetry,
            )
        }
        if (PushHealth.hasUndelivered(snapshot)) {
            UndeliveredRow(
                setAside = snapshot.setAside,
                skipped = snapshot.skipped,
                labelForSourceId = labelForSourceId,
                unitLabelForSourceId = unitLabelForSourceId,
                onDiscard = { confirmingDiscard = true },
            )
        }
    }
    if (confirmingDiscard) {
        DiscardUndeliveredDialog(
            setAside = snapshot.setAside,
            skippedCount = snapshot.skipped.size,
            unitLabelForSourceId = unitLabelForSourceId,
            onConfirm = {
                confirmingDiscard = false
                onDiscardUndelivered()
            },
            onDismiss = { confirmingDiscard = false },
        )
    }
}

@Composable
private fun BlockedRow(blockedSourceIds: List<String>, labelForSourceId: (String) -> String) {
    val c = OmTheme.colors
    val one = blockedSourceIds.size == 1
    Column(verticalArrangement = Arrangement.spacedBy(OmSpacing.xs)) {
        RowTitle("Omnesis isn't accepting some data", Icons.Outlined.Lock, c.danger)
        Text(
            joinNames(blockedSourceIds.map(labelForSourceId)),
            fontSize = 12.sp,
            lineHeight = 17.sp,
            fontWeight = FontWeight.Medium,
            color = c.textPrimary,
        )
        Text(
            "This phone isn't allowed to send ${if (one) "it" else "them"}. Granting that on Omnesis — or " +
                "pairing this phone again — is the fix; retrying isn't. Nothing queued here is thrown away " +
                "meanwhile, but Android expires what ${if (one) "this source reads" else "these sources read"} " +
                "on its own schedule.",
            fontSize = 12.sp,
            lineHeight = 17.sp,
            color = c.textMuted,
        )
    }
}

@Composable
private fun BacklogRow(
    snapshot: PushHealthSnapshot,
    retryPhase: PushHealth.RetryPhase,
    labelForSourceId: (String) -> String,
    unitLabelForSourceId: (String, Int) -> String,
    onRetry: () -> Unit,
) {
    val c = OmTheme.colors
    Column(verticalArrangement = Arrangement.spacedBy(OmSpacing.xs)) {
        RowTitle("Data isn't reaching Omnesis", Icons.Outlined.SyncProblem, c.warning)
        snapshot.queued.forEach { queue ->
            val waited = queue.oldestAgeMillis?.let { ", oldest waiting ${PushHealth.waitedLabel(it)}" }.orEmpty()
            Text(
                "${labelForSourceId(queue.sourceId)} — ${unitLabelForSourceId(queue.sourceId, queue.count)}$waited",
                fontSize = 12.sp,
                lineHeight = 17.sp,
                fontWeight = FontWeight.Medium,
                color = c.textPrimary,
            )
        }
        Text(
            "Omnesis hasn't taken it. It stays on this phone until it does.",
            fontSize = 12.sp,
            lineHeight = 17.sp,
            color = c.textMuted,
        )
        RetryControl(retryPhase, onRetry)
        if (retryPhase is PushHealth.RetryPhase.Reported) {
            RetryStatusLine(retryPhase.outcome)
        }
    }
}

@Composable
private fun RetryControl(retryPhase: PushHealth.RetryPhase, onRetry: () -> Unit) {
    val c = OmTheme.colors
    if (retryPhase is PushHealth.RetryPhase.Running) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(OmSpacing.xs),
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(top = 4.dp),
        ) {
            OmSpinner(modifier = Modifier.size(14.dp), color = c.accent, strokeWidth = 2.dp)
            Text("Syncing…", fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = c.textMuted)
        }
    } else {
        TextButton(onClick = onRetry, contentPadding = androidx.compose.foundation.layout.PaddingValues(0.dp)) {
            Text(
                if (retryPhase is PushHealth.RetryPhase.Idle) "Retry now" else "Try again",
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                color = c.accent,
            )
        }
    }
}

/**
 * The one line a finished retry leaves behind. Its own composable so a golden
 * can render every outcome side by side, which the banner itself only ever
 * shows one at a time.
 */
@Composable
fun RetryStatusLine(outcome: RetryOutcome) {
    val c = OmTheme.colors
    Text(
        PushHealth.retryMessage(outcome),
        fontSize = 12.sp,
        lineHeight = 17.sp,
        color = if (PushHealth.retryIsTrouble(outcome)) c.warning else c.textMuted,
    )
}

@Composable
private fun UndeliveredRow(
    setAside: List<SetAside>,
    skipped: List<SkippedPush>,
    labelForSourceId: (String) -> String,
    unitLabelForSourceId: (String, Int) -> String,
    onDiscard: () -> Unit,
) {
    val c = OmTheme.colors
    Column(verticalArrangement = Arrangement.spacedBy(OmSpacing.xs)) {
        RowTitle("Some data never reached Omnesis", Icons.Outlined.Inbox, c.textPrimary)
        if (setAside.isNotEmpty()) {
            setAside.forEach { held ->
                Text(
                    "${labelForSourceId(held.sourceId)} — ${unitLabelForSourceId(held.sourceId, held.count)} set aside",
                    fontSize = 12.sp,
                    lineHeight = 17.sp,
                    fontWeight = FontWeight.Medium,
                    color = c.textPrimary,
                )
            }
            Text(
                "These were refused often enough that this phone stopped re-sending them, and keeps them " +
                    "as evidence instead. It holds a bounded number, so the oldest are dropped as new ones arrive.",
                fontSize = 12.sp,
                lineHeight = 17.sp,
                color = c.textMuted,
            )
        }
        if (skipped.isNotEmpty()) {
            skipped.forEach { marker ->
                Text(
                    "${labelForSourceId(marker.sourceId)} — ${marker.unit}",
                    fontSize = 12.sp,
                    lineHeight = 17.sp,
                    fontWeight = FontWeight.Medium,
                    color = c.textPrimary,
                )
            }
            Text(
                "Syncing moved past this so everything behind it could keep flowing. Omnesis read it from " +
                    "Android rather than storing a copy here, so it only arrives if the source is removed and " +
                    "added again in Omnesis — and only as far back as Android still keeps it.",
                fontSize = 12.sp,
                lineHeight = 17.sp,
                color = c.textMuted,
            )
        }
        TextButton(onClick = onDiscard, contentPadding = androidx.compose.foundation.layout.PaddingValues(0.dp)) {
            Text("Discard", fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = c.danger)
        }
    }
}

/**
 * Confirms a discard, because it cannot be undone: the retained rows are
 * deleted outright, and dropping the markers removes the only thing on this
 * phone that says a re-sync would be worth running.
 */
@Composable
fun DiscardUndeliveredDialog(
    setAside: List<SetAside>,
    skippedCount: Int,
    unitLabelForSourceId: (String, Int) -> String,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    val c = OmTheme.colors
    val held = setAside.joinToString(" and ") { unitLabelForSourceId(it.sourceId, it.count) }
    val message = buildString {
        if (setAside.isNotEmpty()) {
            append("$held will be deleted from this phone")
            if (skippedCount > 0) append(", and the record of what syncing skipped will be cleared")
            append(". ")
        } else {
            append("The record of what syncing skipped will be cleared. ")
        }
        append("This can't be undone.")
    }
    AlertDialog(
        onDismissRequest = onDismiss,
        containerColor = c.bgSecondary,
        title = { Text("Discard undelivered data?", color = c.textPrimary) },
        text = { Text(message, color = c.textSecondary, fontSize = 13.sp, lineHeight = 19.sp) },
        confirmButton = { TextButton(onClick = onConfirm) { Text("Discard", color = c.danger) } },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Keep", color = c.accent) } },
    )
}

@Composable
private fun RowTitle(text: String, icon: ImageVector, tint: Color) {
    Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
        Icon(
            icon,
            contentDescription = null,
            tint = tint,
            modifier = Modifier.size(14.dp).clearAndSetSemantics {},
        )
        Text(text, fontSize = 13.sp, fontWeight = FontWeight.SemiBold, color = tint)
    }
}

/** "A", "A and B", "A, B and C" — the same shape a person would write. */
internal fun joinNames(names: List<String>): String = when (names.size) {
    0 -> ""
    1 -> names[0]
    else -> names.dropLast(1).joinToString(", ") + " and " + names.last()
}

// MARK: - Previews

private const val HOUR_MILLIS = 60L * 60 * 1000

private val previewLabels: (String) -> String = { id ->
    when (id) {
        "activity-segments:local" -> "Activity Segments"
        "photos:local" -> "Photos"
        else -> "Health Connect"
    }
}

private val previewUnits: (String, Int) -> String = { id, count ->
    val plural = if (id == "photos:local") "photos" else "events"
    val noun = if (count == 1) plural.removeSuffix("s") else plural
    "$count $noun"
}

@Composable
private fun PreviewBanner(
    snapshot: PushHealthSnapshot,
    retryPhase: PushHealth.RetryPhase = PushHealth.RetryPhase.Idle,
    dark: Boolean = true,
) {
    OmnesisTheme(darkTheme = dark) {
        PushHealthBanner(
            snapshot = snapshot,
            retryPhase = retryPhase,
            labelForSourceId = previewLabels,
            unitLabelForSourceId = previewUnits,
            onRetry = {},
            onDiscardUndelivered = {},
            modifier = Modifier.padding(OmSpacing.lg),
        )
    }
}

@Preview(name = "Push health · blocked · dark")
@Composable
private fun PushHealthBlockedPreview() {
    PreviewBanner(PushHealthSnapshot(blockedSourceIds = listOf("photos:local")))
}

@Preview(name = "Push health · backlog · dark")
@Composable
private fun PushHealthBacklogPreview() {
    PreviewBanner(
        PushHealthSnapshot(
            queued = listOf(
                dev.omnesis.android.transport.QueuedBacklog("activity-segments:local", 412, 72 * HOUR_MILLIS),
            ),
        ),
    )
}

@Preview(name = "Push health · undelivered · dark")
@Composable
private fun PushHealthUndeliveredPreview() {
    PreviewBanner(
        PushHealthSnapshot(
            setAside = listOf(SetAside("activity-segments:local", 128)),
            skipped = listOf(
                SkippedPush(
                    sourceId = "photos:local",
                    unit = "42 photos in the backfill pass, added after 2 Jan 2026, 09:00",
                    refusals = 5,
                    atMillis = 0,
                ),
            ),
        ),
    )
}
